include/seastar/util/chunked_deque.hh

```diff
@@ -0,0 +1,733 @@
+chunked_deque<T, items_per_chunk, save_free_chunks>::operator=(chunked_deque&& x) noexcept {
+    if (&x != this) {
+        this->~chunked_deque();
+        new (this) chunked_deque(std::move(x));
+    }
+    return *this;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline bool
+chunked_deque<T, items_per_chunk, save_free_chunks>::empty() const noexcept {
+    return _front_chunk == nullptr;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline size_t
+chunked_deque<T, items_per_chunk, save_free_chunks>::size() const noexcept{
+    if (_front_chunk == nullptr) {
+        return 0;
+    } else if (_back_chunk == _front_chunk) {
+        // Single chunk.
+        return _front_chunk->size();
+    } else {
+        return _front_chunk->size() + _back_chunk->size()
+                + (_nchunks - 2) * items_per_chunk;
+    }
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+void chunked_deque<T, items_per_chunk, save_free_chunks>::clear() noexcept {
+    while (!empty()) {
+        pop_front();
+    }
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks> void
+chunked_deque<T, items_per_chunk, save_free_chunks>::shrink_to_fit() noexcept {
+    while (_free_chunks) {
+        auto next = _free_chunks->next;
+        delete _free_chunks;
+        _free_chunks = next;
+    }
+    _nfree_chunks = 0;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+chunked_deque<T, items_per_chunk, save_free_chunks>::~chunked_deque() {
+    clear();
+    shrink_to_fit();
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+typename chunked_deque<T, items_per_chunk, save_free_chunks>::chunk*
+chunked_deque<T, items_per_chunk, save_free_chunks>::new_chunk() {
+    chunk* ret;
+    if (_free_chunks) {
+        ret = _free_chunks;
+        _free_chunks = _free_chunks->next;
+        --_nfree_chunks;
+    } else {
+        ret = new chunk;
+    }
+    _nchunks++;
+    return ret;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::ensure_room_front() {
+    // If we don't have a back chunk or it's full, we need to create a new one
+    if (_front_chunk != nullptr && _front_chunk->begin > 0) {
+        return;
+    }
+    chunk* n = new_chunk();
+    n->prev = nullptr;
+    if (_front_chunk) {
+        _front_chunk->prev = n;
+        n->begin = n->end = items_per_chunk;
+    } else {
+        _back_chunk = n;
+        n->begin = n->end = items_per_chunk / 2;
+    }
+    n->next = _front_chunk;
+    _front_chunk = n;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::ensure_room_back() {
+    // If we don't have a back chunk or it's full, we need to create a new one
+    if (_back_chunk != nullptr && _back_chunk->end < items_per_chunk) {
```

Comment:
Again I didn't understand the logic why end < items_per_chunk means it is full?

Reply:
since it isn't circular